### ISSUE TITLE
Easier binding of host Kotlin functions to Kotlin/Wasm running within Chasm

### ISSUE THREAD

**Issue by Laxystem** [CharlieTap/chasm#67]
Chasm looks like an incredibly cool project, and the possibilities are countless.

The most natural extension to WASI support would be running containerized Kotlin/WasmWasi using Chasm, with access to specific packages from the host, similarly to KotlinScript (`.kts`).

Therefore, my suggestion—allow to simply reference functions or even entire packages to add them as host functions (and types) for the Chasm container, for example:
```kotlin
fun Foo.myHostApi(): Bar

instance.expose(Foo::myHostApi)
```

My optimal API for this would be to have some kind of `@ChasmExposed` annotation:

```kotlin
instance.exposeAnnotated<ChasmExposed>()
instance.exposePackage("com.example.gameEngine")
```

Reflection data (that is not available in runtime due to multiplatform limitations) could be made up for using some kind of compiler plugin (discovering packages and annotations for this purpose should probably be done in compile-time anyway).

But of course, the initial request is to simply be able to expose functions automagically.

The ultimate extension to this would be to somehow support (an improved version of) the KotlinScript API, but this is a separate issue entirely that'll likely need to be developed in collaboration with the Kotlin Foundation.

**Comment by Laxystem** [comment]
I'll expend on my use case: the ability to have pure, scripty Kotlin, without *any* form of access to implementation details or the running OS.

KotlinScript lacks in this, as it can still bypass `@RequiresOptIn` APIs, and it was very hard if not impossible to containerize it.

For my game engine, I'm looking for a way to simply have, say:

```kotlin
// example.system.kt

for ((entity, gravity) in entities[Gravity]) entity[Position] += gravity.modifier * deltaTime
```

Compile it into Kotlin/Wasm, and run it *anywhere*, without any fear of platform compatibility, security, anything—just containerized logic that only has access to what I give it. KotlinScript, but good, that is. So here I am, suggesting the first step towards it: adding Kotlin functions to the container without unnecessary boilerplate.

Then, it'd be types, then, automatic discovery of some sort, then, KotlinScript syntax, and walla!

**Comment by CharlieTap** [comment]
Hey @Laxystem 

This is something thats planned, a while back I added a [moduleInfo api](https://github.com/CharlieTap/chasm/blob/main/chasm/src/commonMain/kotlin/io/github/charlietap/chasm/embedding/ModuleInfo.kt) which describes a wasm module in terms of its imports and exports alongside all the necessary type information. The idea is that you can easily use this information inside of a gradle plugin to generate a kotlin interface and implementation which hides chasm and its low level api. This is the end goal of how I want people to interact with chasm because it would make end users less dependent on chasms abi, and thus could innovate without worrying about breaking peoples programs.

TLDR its high on the list but not just yet, I'm open to prs and also the api to make this happen is public so you could even have a distinct library that just wraps chasm if you wanted
